### ISSUE TITLE
e_sqlite3.dll should not be copied to the publish output

### ISSUE THREAD

**Issue by JunTaoLuo** [ericsink/SQLitePCL.raw#154]
Following up to #140, this dll should not be put into the publish output. This can be achieved by adding the `CopyToPublishDirectory` property and setting it to `Never` i.e.
```xml
      <Content Include="$(MSBuildThisFileDirectory)..\runtimes\win7-x86\native\e_sqlite3.dll">
        <Link>x86\e_sqlite3.dll</Link>
        <CopyToOutputDirectory>PreserveNewest</CopyToOutputDirectory>
        <Pack>false</Pack>
        <CopyToPublishDirectory>Never</CopyToPublishDirectory>
      </Content>
      <Content Include="$(MSBuildThisFileDirectory)..\runtimes\win7-x64\native\e_sqlite3.dll">
        <Link>x64\e_sqlite3.dll</Link>
        <CopyToOutputDirectory>PreserveNewest</CopyToOutputDirectory>
        <Pack>false</Pack>
        <CopyToPublishDirectory>Never</CopyToPublishDirectory>
      </Content>
```

Note that this target is only needed for apps targeting .NET Framework and does not apply to apps targeting .NET Core. As such the target file should be moved to a subdirectory specific to .NET Framework:
```
nupkg_root
\_ build
  \_ net45 (or the lowest supported .NET Framework version)
    \_ SQLitePCLRaw.lib.e_sqlite3.v110_xp.targets
\_ SQLitePCLRaw.lib.e_sqlite3.v110_xp.nuspec
\_ ... other files
```
This will probably also require a `<dependencies>Any</dependencies>` attribute in the `.nuspec` file.  @natemcmaster has more context on this.

**Comment by JunTaoLuo** [comment]
cc @ericsink

**Comment by ericsink** [comment]
So this is actually a different issue from #140, which dealt with whether the DLL ended up in nuget packaging.  This issue is something that hasn't been brought up before, AFAIK.

Can you by chance point me to any docs for CopyToPublishDirectory?  Googling that term mostly results in people complaining that something doesn't work, rather than explanations of how it *should* work.  :-)  

Do I need to worry that CopyToPublishDirectory is used in some context where CopyToPublishDirectory=Never is not the right answer for e_sqlite3.dll.  I suppose if this "publish" property is never used outside of web projects, that would make sense.

**Comment by JunTaoLuo** [comment]
Right, this is a different issue but has similar symptoms. I don't know if there is any documentation on the attribute @pranavkm @natemcmaster? However, the source for the publish targets are here: https://github.com/dotnet/sdk/blob/39102ec214514363c6af4cd9d599a35aceb8c745/src/Tasks/Microsoft.NET.Build.Tasks/build/Microsoft.NET.Publish.targets

**Comment by ericsink** [comment]
I am proceeding with `<CopyToPublishDirectory>Never</CopyToPublishDirectory>` assuming it to be safe.

In the form of an intentionally dumb question, what I'm looking for is:  What is the publish directory, and why should `e_sqlite3.dll` not be copied to it?

**Comment by natemcmaster** [comment]
Actually, I'm missing something here too. @JunTaoLuo why do we not want `e_sqlite3.dll` in the publish folder? IIUC for .NET Framework, we want `e_sqlite3.dll` in the `$(PublishDir)/x64/` and  `$(PublishDir)/x86/` folders so AnyCPU projects will run.

**Comment by natemcmaster** [comment]
@ericsink CopyToPublishDirectory = {Never, PreserveNewest, Always}. It's new to Microsoft.NET.SDK AFAIK.

**Comment by JunTaoLuo** [comment]
The intention is that these binaries will be included in the Runtime Package Store and thus should be excluded from the publish output directory.

**Comment by natemcmaster** [comment]
Oh, I see. Ignoring the Runtime Package Store for a minute, the gist of the problem that .NET Core apps end up with `e_sqlite3.dll` in published output, right? One copy in `$(PublishDir)/runtimes/win7-{x64,x86}/e_sqlite3.dll` and another in `$(PublishDir)/{x64,x86}/e_sqlite3.dll`

**Comment by ericsink** [comment]
I'm more confused now, as I thought this issue was specific to .NET Framework, and unrelated to .NET Core, but the latest comment from @natemcmaster seems to contradict that.

I'm also still (and perhaps more) concerned that the proposed change is to get behavior which could be considered specific to EF Core.  But I'm not at all sure.

**Comment by JunTaoLuo** [comment]
Yes, currently publishing a .NET Core app results in the e_sqlite3.dll appearing in both locations. The expected behavior is to only have it in `$(PublishDir)/runtimes/win7-{x64,x86}/e_sqlite3.dll`. 

I believe the target was added originally to solve the problem of native assemblies on .NET Framework. However, since the target was placed in the `$(NupkgRoot)/build`, the customized targets were being picked up for both .NET Framework and .NET Core apps leading to the duplication of the native assets on .NET Core. 

Now I think about it, I think it would be sufficient to just place the target under `$(NupkgRoot)/build/net45` or whatever the minimum supported .NET Framework version is. Do you agree @natemcmaster ?

**Comment by ericsink** [comment]
OK, I think the fog in my brain has cleared and I get this now.

Moving the .targets file does look like the right thing to do.  And if I understand correctly, the Linux and MacOS packages will need the same fix for .NET Core support on those platforms.

**Comment by natemcmaster** [comment]
Right. NuGet handles the 'runtimes/$rid/native/' folder automatically for .NET Core projects. It also handles it for .NET Framework projects building with PackageReference (not packages.config).

The only reason you should need .targets files to copy native .dll's are:

- .NET Framework with packages.config (NuGet v2)
- .NET Framework that target AnyCPU. cc @nguerrera has been doing work in this area. See https://github.com/dotnet/sdk/issues/843

So, I'd echo the proposed changes John mentioned above, which are 

1. Remove the .targets files for macOS and Linux packages (unless you are trying to support Mono...which is a whole different can of worms.)
2. Don't rely on `"$(OS)"` to determine if you are building for windows. With the .NET Core CLI, you can build a Windows app on Linux and a Linux app on Windows. `$(OS)` determines the the OS running MSBuild, `$(RuntimeIdentifier)` determines which OS the app targets.
3. Move SQLitePCLRaw.lib.e_sqlite3.v110_xp.targets to `build/net45/`


Also, you might consider changing the .targets file to something like this: 
```xml
<Project xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
<ItemGroup>
  <None Include="$(MSBuildThisFileDirectory)..\runtimes\win7-x86\native\e_sqlite3.dll">
    <Link>x86\e_sqlite3.dll</Link>
    <CopyToOutputDirectory>PreserveNewest</CopyToOutputDirectory>
    <CopyToPublishDirectory>PreserveNewest</CopyToPublishDirectory>
    <Visible>false</Visible>
  </None>
  <None Include="$(MSBuildThisFileDirectory)..\runtimes\win7-x64\native\e_sqlite3.dll">
    <Link>x64\e_sqlite3.dll</Link>
    <CopyToOutputDirectory>PreserveNewest</CopyToOutputDirectory>
    <CopyToPublishDirectory>PreserveNewest</CopyToPublishDirectory>
    <Visible>false</Visible>
  </None>
</ItemGroup>
</Project>
```
This has the advantage of not relying on target chaining (a common source of failures). `None` items do not end up in NuGet packages (i.e. Pack defaults to true on Content, false on None). Visible=false hides the file from Solution Explorer.

**Comment by ericsink** [comment]
That's good info, thanks.

I do need to keep support for Mono though...

**Comment by natemcmaster** [comment]
Ok. Then, keep the targets. Mono is treated as .NET Framework, so `build/net45` should work there too.

After thought: because the \*.targets file is in `build/`, NuGet regards the package as compatible with any framework. By moving it to `build/net45/` NuGet may regard the package as only compatible with `net45` and up. As @JunTaoLuo mentioned, you can probably workaround this by adding frameworks to you nuspec. i.e.
```xml
<package>
  <metadata>
    <!-- ... -->
    <dependencies>
    	<group targetFramework=".NETStandard1.0" />
    	<group targetFramework=".NETFramework4.5" />
    </dependencies>
  </metadata>
</package>
```

**Comment by JunTaoLuo** [comment]
Wondering when will we be able to get a new version of this package with the aforementioned fix?

**Comment by ericsink** [comment]
Maybe today.  Hopefully Monday worst case.  I was testing yesterday to make sure I didn't break anything.

**Comment by ericsink** [comment]
The main issue here (mv the .targets files down into build/net35 so they won't affect .NET Core) is fixed in 1.1.4, which I have pushed to nuget.

The other comments about the .targets file still need attention, and have been logged as a separate issue, #156.

**Comment by natemcmaster** [comment]
Thanks Eric!
